### ISSUE TITLE
[Feature] Test Plugins

### ISSUE THREAD

**Issue by rwoll** [microsoft/playwright#13362]
Hook into Playwright Test's lifecycle to customize behavior easily, with minimal config. Similar to Custom Reporter, but spans entire run lifecycle, including modifying state (pushing attachments, influencing sharding/scheduling decisions, adding CLI flags, etc.).

Some uses for plugins:
- generating tests / data for use in tests (e.g. yaml file -> test cases run via Playwright, Gherkin style tests)
- test compilation and/or transformation (e.g. custom tsx)
- scheduling/sharding: currently, PWT's scheduling and sharding logic is static
- set multiple configurations at once (e.g. worker count, reporter, connection, webServers/launchServices)
- run test on file change
- dynamically set timeouts
- add default fixtures (like `test.extend`)

Some quick ideas for API:

```ts
import { webServer } from '@rwoll/playwright-plugins';
import { PlaywrightTestConfig } from '@playwright/test';
const config: PlaywrightTestConfig = {
  plugins: [
     compile('./src'),
     webServer('./html/files'),
     remoteConnection(),
     testsFromDB(dbConnectionString),
     // …
  ],
};
export default config;
```

```ts
interface Plugin {
    setup(config);
    collectFiles();
    schedule();
    …
}
```

Some open questions?

- how do plugins interact with one another? For example, if we run the `webServer` plugin, can it output a port/URL for other plugins to consume? Can we get typesafety and dependency ordering by make plugins like fixtures?

**Comment by ryanjduffy** [comment]
I think this would be very useful for our replay.io integration. We've added some integration via [action-playwright](https://github.com/Replayio/action-playwright) and a [custom reporter](https://github.com/Replayio/replay-cli/tree/main/packages/playwright) but we need to go one step deeper to be able to set environment variables for the browser for each worker.

I've explored using a fixture for this which I think can work but seems to require the consumers explicitly use the fixture from their tests which is not ideal. I'd rather rely on declarative config so the test files stay the same but they can opt in via config to using the Replay browser.
